src/Lykke.Service.ExchangeConnector/Handlers/OrderBookHandlerDecorator.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using TradingBot.Trading;
+using TradingOrderBook = TradingBot.Trading.OrderBook;
+using LykkeOrderBook = TradingBot.Exchanges.Concrete.LykkeExchange.Entities.OrderBook;
+
+namespace TradingBot.Handlers
+{
+    internal class OrderBookHandlerDecorator : IHandler<LykkeOrderBook>
+    {
+        public static readonly string Name = "lykke";
+        private readonly IHandler<TradingOrderBook> _rabbitMqHandler;
+        private readonly Dictionary<string, LykkeOrderBook> halfOrderBooks = new Dictionary<string, LykkeOrderBook>();
+
+        public OrderBookHandlerDecorator(IHandler<TradingOrderBook> rabbitMqHandler)
+        {
+            _rabbitMqHandler = rabbitMqHandler;
+        }
+
+        public async Task Handle(LykkeOrderBook message)
+        {
+            // Find a pair and send it
+            var wantedKey = message.AssetPair + !message.IsBuy;
+            halfOrderBooks.TryGetValue(wantedKey, out var otherHalf);
+            if (otherHalf != null)
+            {
+                var fullOrderBook = CreateOrderBook(Name, message, otherHalf);
+                await _rabbitMqHandler.Handle(fullOrderBook);
+            }
+
+            // Update current order book
+            var currentKey = message.AssetPair + message.IsBuy;
```

Comment:
I would move adding code above publishing. In case of an exception in publishing the snapshot will stay actual